FIX: Post height issues with Expand Table button

keegangeorge opened the pull request:
This PR aims to fix _post height issues_ with the wrapper that adds the <kbd>Expand Table</kbd> button (and <kbd>Edit Table</kbd> button added by [`discourse-table-builder`](https://github.com/discourse/discourse-table-builder), which is has a simultaneous fix [here](https://github.com/discourse/discourse-table-builder/pull/34)).

It does so by ensuring the table wrapper buttons are present inside a container, which is overlaid on-top of the table wrapper on hover, rather than being inline with the post content.

Having the table wrapper buttons inline with the post content was causing a gap between text and tables.

tgxworld commented:
Are the changes here testable? I think it is important especially when the changes we are making here are meant to accommodate external plugins.

keegangeorge commented:
> Are the changes here testable? I think it is important especially when the changes we are making here are meant to accommodate external plugins.

These changes are quite difficult to test. In short we are merely changing it from being:

```handlebars
<div> <-- md-table-wrapper
<button></button> <-- Expand button
<table></table> <-- table
</div>
```

to 

```handlebars
<div> <-- md-table-wrapper
<div><button></button></div> <-- wrapper for buttons overlaid on tables
<table></table> <-- table
</div>
```

tgxworld commented:
> These changes are quite difficult to test

Can you clarify why it would be difficult to test? If it is the HTML structure we care about, it should be easily testable with an acceptance test.

> In short we are merely changing it from being:

Instead of thinking of the change as small as the deciding factor whether we need a test, I think it'll be more beneficial if we think about what implication does this change have if it regressed without us knowing about it. In this case, I wonder if it'll be helpful for us to have a test which ensures that the buttons are not inlined with the post's content which is known to cause problems.

keegangeorge commented:
Thanks @tgxworld, I appreciate the feedback! I noticed there weren't any tests for the Expand Table at all, so I've added some tests along with the test to ensure post height issues don't arise. I've also built a pluginAPI method so extensions (such as the table builder) can easily add buttons onto the table wrapper without duplicating any logic.

I'm running into a small issue though with the implementation. It occurs topics/posts with more than one table, where the button added using the new api method only appears on the last table element. I'm still looking into resolving this, so I'm marking this PR as draft. If you have any ideas on why its happening I'd appreciate the help :)

Note, I've created a test topic with the markdown:
```markdown
|Make | Model | Year | Price | Drivetrain | Engine | Power | Type | Seats | Wheels | Mileage | Induction | Intake | Ranking|
|--- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | ---|
|Toyota | Supra | 1998 | 50k | RWD | I6 | 400HP | Coupe | 4 | Enkei PF01 | 100,000 | Twin Turbo | Yes | 1|
|Nissan | Skyline | 1999 | 100k | 4WD | I6 | 350HP | Coupe | 4 | TE37 | 80,000 | Single Turbo | No | 2|
|Honda | S2000 | 2001 | 30k | RWD | I4 | 230HP | Roadster | 2 | RPF1 | 110,000 | Naturally Aspirated | No | 3|


|Make | Model | Year|
|--- | --- | ---|
|Toyota | Corolla | 1998|
|Nissan | Altima | 1999|
|Honda | Civic | 2001|
```

and what's happening is the test button (labelled "Draft") that I've added only is appearing on the second table...

tgxworld commented on app/assets/javascripts/discourse/app/lib/plugin-api.js:
```diff
@@ -2014,6 +2018,31 @@ class PluginApi {
+   * @param {function} event - A function that returns an event to be dispatched when the button is clicked
+   *
+   */
+  addTableWrapperButton(label, icon, classes, event) {
```
Just looking at this from an API perspective, is there a way to target a particular table in a post? It seems like this API will current just add the button to all the tables in the post.

keegangeorge replied:
Yes, it should be possible through the `event` by getting the element's `data-table-id` attribute like so:
```javascript
const tableId = event.target.getAttribute("data-table-id");
```

It's not immediately, obvious, so I'll be sure to add that to the comment.

tgxworld commented:
One small suggestion I have is to ship your fix and the new plugin API separately since they can be reviewed and shipped independently.

keegangeorge commented:
@tgxworld Sounds good, I've moved the pluginAPI related changes to this [PR](https://github.com/discourse/discourse/pull/18871) and kept this PR strictly to fix the post height issue.

pmusaraj commented on app/assets/stylesheets/common/base/topic-post.scss:
```diff
@@ -1633,20 +1633,21 @@ span.mention {
+  &-buttons {
+    position: absolute;
+    top: 0.5rem;
+    left: 0.5rem;
```
<img width="719" alt="image" src="https://user-images.githubusercontent.com/368961/227594650-a3c78392-c5aa-4a69-8975-cb7bc941d847.png">

I'm ok with this, but since we're now overlaying these buttons, I wonder whether in a followup PR we should switch to using icon-only buttons. It would take up a lot less space (and I think most users can still easily tell what the buttons do by just seeing icons).

pmusaraj commented on app/assets/javascripts/discourse/app/initializers/post-decorations.js:
```diff
@@ -176,8 +177,12 @@ export default {
-          table.parentNode.insertBefore(popupBtn, table);
+          // Create a button wrapper for case of multiple buttons (i.e. table builder extension)
+          const buttonWrapper = document.createElement("div");
+          buttonWrapper.classList.add("fullscreen-table-wrapper-buttons");
```
```suggestion
          buttonWrapper.classList.add("fullscreen-table-wrapper--buttons");
```

Super-nitpicky, but slightly more BEMy class name with the two dashes.
